programs/survival/compass.sc

```diff
@@ -0,0 +1,613 @@
+    slots = l(); // The list of the items that need to be given back when the player respawns
+
+    // Find the slots of all the compasses
+    slot = inventory_find(player, 'compass');
+    while (slot != null, 41,
+        // Only add the compass to the list if it's a tracking compass
+        if (inventory_get(player, slot): 2: 'Tracking',
+            triple = inventory_get(player, slot);
+            put(slots, null, l(triple: 0, triple: 1, triple: 2, slot));
+            inventory_set(player, slot, 0);
+        );
+        slot = inventory_find(player, 'compass', slot+1);
+    );
+
+    bySlotNum = map(slots, _: 3); // All the inventory slots that are taken
+
+    // Find all the slots where the player has items that they respawn with
+    for (global_toKeep,
+        if (_: 3 == 'âny' || _: 3 == lower(player ~ 'name'), // It's impossible to use the character ̂  in a player name
+            // Find the first slot with the item
+            slot = inventory_find(player, _: 0);
+
+            if (slot == null,
+                // If the slot doesn't exist, pick a slot later so it won't be overwritten by other items
+                put(slots, null, l(copy(_: 0), copy(_: 1), copy(_: 2), 'findLater'));
+            ,
+                // If the slot does exist, save it in slots and note that the slot is now taken
+                triple = inventory_get(player, slot);
+                put(slots, null, l(triple: 0, copy(_: 1), copy(_: 2), slot));
+                inventory_set(player, slot, 0);
+                put(bySlotNum, null, slot);
+            );
+        );
+    );
+
+
+    // For all the items that aren't in the player's inventory when they die, pick the first spot that isn't taken
+    for (slots,
+        if (_: 3 == 'findLater',
+            testSlot = 0;
+
+            while (includes(bySlotNum, testSlot), 41, testSlot = testSlot+1);
+
+            _: 3 = testSlot;
+            put(bySlotNum, null, testSlot);
+        );
+    );
+
+    // Save the slots in a variable for when they respawn
+    put(global_inventoryData, player ~ 'name', slots);
+);
+
+__on_player_respawns(playerOof)->(
+    schedule(0, _(outer(playerOof)) -> (
+        player = player(playerOof ~ 'name');
+        slots = global_inventoryData: (player ~ 'name');
+
+        // When the player respawns, give them the items that are saved in the inventory spots they were when the player died, essentially keepInventory but for certain items
+        for (slots,
+            inventory_set(player, _: 3, _: 1, _: 0, _: 2);
+        );
+
+        // Apply effects
+        for (global_toKeepEffect,
+            if (_: 3 == 'âny' || lower(_: 3) == player ~ 'name',
+                modify(player, 'effect', _: 0, _: 1, _: 2, false);
+            );
+        );
+
+    );
+    );
+);
+
+global_playerPos = {};
+
+changeCompassPos() -> {
+    // Store the current locations of all the players
+    updatePlayerPosition();
+
+    players = entity_selector('@a');
```

Comment:
Why do you sometime use entity_selector('@a') and sometimes player('*')?

Reply:
The only time I could see `player('*')` being used was the part of debug.sc that ghoulboy wrote and I modified a bit, [linked here](https://github.com/gnembon/scarpet/pull/177#discussion_r575848600). I use `entity_selector('@a')` because it gets players from every dimension, which is usually what I want

Reply:
My bad, the correct usage would be `player('all')` to get all players from all dimensions in the fastest way possible. `entity_selector('@a')` obviously works, but is slower, as it has to go through the entire rigmarole of the vanilla entity querying thing ma bob, then returns the players, whereas scarpet just reads the contents of an internal Java variable.